src/slice.c

```diff
@@ -100,98 +103,136 @@ static SEXP vec_slice_fallback(SEXP x, SEXP index) {
+
+  if (nms == R_NilValue) {
+    UNPROTECT(1);
+    return;
+  }
+
+  nms = PROTECT(chr_slice(nms, index));
+  Rf_setAttrib(x, R_NamesSymbol, nms);
+
+  UNPROTECT(2);
+}
+
+/**
+ * @param to The type to restore to. Necessary to pass it along
+ *   because `vec_slice_native()` passes a proxy but needs to restore
+ *   to the original type.
+ */
+static SEXP vec_slice_impl(SEXP x, SEXP index, SEXP to, bool dispatch) {
   if (has_dim(x)) {
     return vec_slice_fallback(x, index);
   }
 
   SEXP out = R_NilValue;
+  enum vctrs_type type = vec_typeof_impl(x, dispatch);
 
-  switch (vec_typeof_impl(x, dispatch)) {
+  switch (type) {
   case vctrs_type_null:
     Rf_error("Internal error: Unexpected `NULL` in `vec_slice_impl()`.");
 
   case vctrs_type_logical:
-    out = lgl_slice(x, index);
-    break;
   case vctrs_type_integer:
-    out = int_slice(x, index);
-    break;
   case vctrs_type_double:
-    out = dbl_slice(x, index);
-    break;
   case vctrs_type_complex:
-    out = cpl_slice(x, index);
-    break;
   case vctrs_type_character:
-    out = chr_slice(x, index);
-    break;
   case vctrs_type_raw:
-    out = raw_slice(x, index);
-    break;
   case vctrs_type_list:
-    out = list_slice(x, index);
-    break;
+    return vec_slice_base(type, x, index, to, dispatch);
 
-  case vctrs_type_dataframe:
-    out = PROTECT(rows_slice(x, index));
-    out = df_restore(out, x, index);
+  case vctrs_type_dataframe: {
+    SEXP out = PROTECT(df_slice(x, index));
+    out = vec_restore(out, to, index);
     UNPROTECT(1);
     return out;
+  }
 
   case vctrs_type_s3: {
-    SEXP proxy = PROTECT(vec_proxy(x));
-
-    out = PROTECT(vec_slice_impl(proxy, index, false));
-    out = vec_restore(out, x, index);
-
-    UNPROTECT(2);
+    // Normally we'd take the proxy and recurse with dispatch turned off.
+    // However we fall back to `[` for compatibility with foreign types.
+    SEXP call = PROTECT(Rf_lang3(fns_bracket, x, index));
+    out = Rf_eval(call, R_GlobalEnv);
+    UNPROTECT(1);
     return out;
   }
 
   default:
     Rf_error("Internal error: Unexpected type `%s` for vector proxy in `vec_slice()`",
              vec_type_as_str(vec_typeof(x)));
   }
-
-  PROTECT(out);
-  slice_names(out, x, index);
-
-  out = vec_restore(out, x, index);
-
-  UNPROTECT(1);
-  return out;
-}
-
-static void slice_names(SEXP x, SEXP to, SEXP index) {
-  SEXP nms = PROTECT(Rf_getAttrib(to, R_NamesSymbol));
-
-  if (nms == R_NilValue) {
-    UNPROTECT(1);
-    return;
-  }
-
-  nms = PROTECT(chr_slice(nms, index));
-  Rf_setAttrib(x, R_NamesSymbol, nms);
-
-  UNPROTECT(2);
 }
 
-SEXP vctrs_slice(SEXP x, SEXP index, SEXP dispatch) {
+// [[export]]
+SEXP vctrs_slice(SEXP x, SEXP index, SEXP native) {
   if (x == R_NilValue || index == R_MissingArg) {
     return x;
   }
 
   index = PROTECT(vec_as_index(index, x));
-  SEXP out = vec_slice_impl(x, index, *LOGICAL(dispatch));
+
+  SEXP out;
+  if (*LOGICAL(native)) {
```

Comment:
I think `LOGICAL(native)[0]` is a idiom that's slightly easier to understand (and more common)